File: [O]math/[O]math/67-add-binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        res = ''
        carry = 0
        i = len(a) - 1
        j = len(b) - 1
        while carry or i >= 0 or j >= 0:
            if i >= 0:
                if a[i] == '1':
                    carry += 1
                i -= 1
            if j >= 0:
                if b[j] == '1':
                    carry += 1
                j -= 1
            carry, remainder = divmod(carry, 2)
            if remainder == 1:
                res = '1' + res
            else:
                res = '0' + res
        return res    
            
# time O(max(m, n))
# space O(max(m, n))
# using math

File: [O]math/[O]math/67-add-binary.java
class Solution {
    public String addBinary(String a, String b) {
        int i = a.length() - 1;
        int j = b.length() - 1;
        int carry = 0;
        StringBuffer s = new StringBuffer();
        while (carry > 0 || i >= 0 || j >= 0) {
            if (i >= 0) {
                if (a.charAt(i) == '1') {
                    carry += 1;
                }
                i -= 1;
            }
            if (j >= 0) {
                if (b.charAt(j) == '1') {
                    carry += 1;
                }
                j -= 1;
            }
            int quo = carry / 2;
            int rem = carry % 2;
            if (rem == 1) {
                s.append('1');
            } else {
                s.append('0');
            }
            carry = quo;
        }
        return s.reverse().toString();
    }
}

// time O(max(m, n))
// space O(max(m, n))
// using math